File: java/leetcode/src/main/java/com/leetcode/tree/Trim_a_Binary_Search_Tree_669.java
package com.leetcode.tree;

public class Trim_a_Binary_Search_Tree_669 {
    public TreeNode trimBST(TreeNode root, int low, int high) {
        if (root == null) return null;
        TreeNode left = trimBST(root.left, low, high);
        TreeNode right = trimBST(root.right, low, high);
        if (!inRange(root.val, low, high)) {
            if (left == null && right == null) return null;
            else if (left != null && right == null) return left;
            else if(left == null && right != null) return right;
            else {
                // nothing
            }
        }
        root.left = left;
        root.right = right;
        return root;
    }

    boolean inRange(int val, int low, int high) {
        return val >= low && val <= high;
    }
}

File: js/algrithm/tree/Trim_a_Binary_Search_Tree_669.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} low
 * @param {number} high
 * @return {TreeNode}
 */
var trimBST = function(root, low, high) {
    if (root === null) return null
    if (root.val > high) {
        return trimBST(root.left, low, high)
    } else if (root.val < low) {
        return trimBST(root.right, low, high)
    } else {
        root.left = trimBST(root.left, low, root.val)
        root.right = trimBST(root.right, root.val, high)
    }
    return root
};